worker/upload.go

```diff
@@ -0,0 +1,836 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+// TODO: add priority queues
+
+const defaultAvgUpdateInterval = 10 * time.Second
+
+var errNoQueue = errors.New("no queue")
+
+type (
+	uploadID [8]byte
+
+	uploader struct {
+		w *worker
+
+		statsOverdrive *average
+		statsStopChan  chan struct{}
+
+		poolMu sync.Mutex
+		pool   []*uploadQueue
+		bh     uint64
+	}
+
+	uploadQueue struct {
+		u  *uploader
+		hp hostProvider
+		rl revisionLocker
```

Comment:
It seems a bit odd that a queue which usually is a data structure would need all these things.
Imo the queue shouldn't do any uploading but the `uploader` should have a goroutine that pulls from queues.

That also gets rid of the circular dependency you got here where the uploader has queues and the queue has an uploader.